src/modules/assistants/library/AssistantTemplateCard.tsx

```diff
@@ -0,0 +1,142 @@
+ */
+
+import { ToolReference, ToolsUsage } from '@/app/api/tools/types';
+import { CardsListItem } from '@/components/CardsList/CardsListItem';
+import { useVectorStore } from '@/modules/knowledge/hooks/useVectorStore';
+import { useToolInfo } from '@/modules/tools/hooks/useToolInfo';
+import { getToolReferenceFromToolUsage } from '@/modules/tools/utils';
+import { SkeletonText } from '@carbon/react';
+import clsx from 'clsx';
+import {
+  Dispatch,
+  MouseEventHandler,
+  SetStateAction,
+  useCallback,
+  useEffect,
+  useState,
+} from 'react';
+import { AssistantIcon } from '../icons/AssistantIcon';
+import { Assistant, AssistantTemplate } from '../types';
+import classes from './AssistantCard.module.scss';
+
+interface Props {
+  template: AssistantTemplate;
+  selected?: boolean;
+  onClick?: MouseEventHandler;
+  onDeleteSuccess?: (assistant: Assistant) => void;
+}
+
+export function AssistantTemplateCard({ template, selected, onClick }: Props) {
+  const { name, description, tools, tool_resources } = template;
+  const vectorStoreId = tool_resources?.file_search?.vector_store_ids?.[0];
+
+  return (
+    <CardsListItem
+      className={classes.root}
+      title={name ?? ''}
+      icon={<AssistantIcon assistant={template} size="lg" />}
+      onClick={onClick}
+      selected={selected}
+      canHover
+    >
+      {description && (
+        <div className={classes.body}>
+          <p>{description}</p>
+        </div>
+      )}
+
+      <div className={classes.footer}>
+        <ToolsInfo tools={tools} />
+        {vectorStoreId && <KnowledgeInfo vectorStoreId={vectorStoreId} />}
+      </div>
+    </CardsListItem>
+  );
+}
+
+function ToolsInfo({ tools }: { tools: ToolsUsage }) {
+  const displayedTools = tools.slice(0, 3);
+  const remainingToolsCount = tools.length - displayedTools.length;
+  const [loadingStates, setLoadingStates] = useState(
+    Array.from({ length: displayedTools.length }, () => true),
+  );
+  const isLoading = loadingStates.some((state) => state);
+
+  return tools.length > 0 ? (
+    <>
+      {isLoading && <SkeletonText className={classes.info} />}
+
+      <div className={clsx(classes.info, { [classes.loading]: isLoading })}>
+        <strong>Tools:</strong>
+        {displayedTools.map((tool, index) => {
+          const toolReference = getToolReferenceFromToolUsage(tool);
+
+          return (
+            <span key={toolReference.id}>
+              <ToolName
+                index={index}
+                tool={toolReference}
+                setLoadingStates={setLoadingStates}
+              />
+              {index < displayedTools.length - 1 ? ', ' : ''}
+            </span>
+          );
+        })}
+        {remainingToolsCount > 0 && ` +${remainingToolsCount}`}
+      </div>
+    </>
+  ) : null;
+}
+
+function ToolName({
+  index,
+  tool,
+  setLoadingStates,
+}: {
+  index: number;
+  tool: ToolReference;
+  setLoadingStates: Dispatch<SetStateAction<boolean[]>>;
```

Comment:
I would do this more generic like `onLoadingStateChange: () => boolean` and lift the setLoadingStates logic to the parent.
I created the same component just today (`ToolName`), co we could also squash them easily then.

Reply:
I tried that and run into some "Maximum call stack exceeded" error, which I was not able to easily get rid of. Let's merge it now and we can rewrite later, if needed.